src/libmongoc/src/mongoc/mongoc-topology-scanner.c

```diff
@@ -1324,3 +1346,15 @@ _jumpstart_other_acmds (mongoc_topology_scanner_node_t *node,
       }
    }
 }
+
+void
+_mongoc_topology_scanner_set_server_api (mongoc_topology_scanner_t *ts,
+                                         const mongoc_server_api_t *api)
+{
+   BSON_ASSERT (ts);
+   BSON_ASSERT (api);
+
+   mongoc_server_api_destroy (ts->api);
+   ts->api = mongoc_server_api_copy (api);
+   _reset_ismaster (ts);
```

Comment:
When creating a new client using `mongoc_client_new`, this implicitly also creates the `topology_scanner_t` through `mongoc_topology_new` and `mongoc_topology_scanner_new`. This last method prepares the basic `isMaster`/`hello` command. At that point in time, the user hasn't had a chance to call `mongoc_client_set_server_api`, hence the need to reset any previously created handshake commands to recreate them with the versioned API options.

Since `_mongoc_topology_scanner_set_server_api` is part of the private API, I haven't bothered to add checks against changing the previously declared API version, especially since the logic required due to the initialisation flow (i.e. changing from no declared API to a specific API) completely covers the scenario of changing from one declared API version to another.

Reply:
Got it. Thanks for the explanation.